lib/internal/streams/readable.js

```diff
@@ -1287,6 +1361,230 @@ ObjectDefineProperties(ReadableState.prototype, {
   }
 });
 
+function isLocked(self) {
+  return self[kConsume]?.stream?.locked === true;
+}
+
+// https://streams.spec.whatwg.org/#readablestream-disturbed
+function isDisturbed(self) {
+  const { _readableState: state } = self;
+  return !!(
+    state.dataEmitted ||
+    state.endEmitted ||
+    state.errorEmitted ||
+    state.closeEmitted
+  );
+}
+
+// https://fetch.spec.whatwg.org/#body-unusable
+function isUnusable(self) {
+  return isDisturbed(self) || isLocked(self);
+}
+
+async function consume(parent, type) {
+  if (isUnusable(parent)) {
+    // eslint-disable-next-line no-restricted-syntax
+    throw new TypeError('unusable');
+  }
+
+  if (parent[kConsume]) {
+    // TODO: Should multiple consume in same tick be possible?
+    throw new TypeError('unusable');
```

Comment:
Is anyone able to determine whether or not this is spec compliant? https://fetch.spec.whatwg.org/#concept-body-consume-body

My guess is that consume body will lock the resource, hence it cannot be consumed again.

Reply:
Well, my understanding is:

 - If the body is null no type error is thrown and consume body returns a promise fulfilled with the empty byte sequence.
 - If it's not null "fully reading body as promise"  which presumably locks the stream to a reader.

That said - since this is spec territory it's better to ping the knowledgeable folk who worked on this like @domenic and @jakearchibald

Reply:
> My guess is that consume body will lock the resource, hence it cannot be consumed again.

That's my understanding as well. There are a few issues here to think about. Should it be possible, for instance, to consume the stream as a `ReadableStream` if the stream is also being consumed using the event API? It's a fairly significant grey area.

Reply:
> Should it be possible, for instance, to consume the stream as a ReadableStream if the stream is also being consumed using the event API? It's a fairly significant grey area.

I think not. But with this PR the user can choose, `Readable.body` will not allow it while `Readable.toWeb` will allow it.

Reply:
@domenic

Reply:
Sorry for the slow reply!

> My guess is that consume body will lock the resource, hence it cannot be consumed again.

That's true. Following the spec, from:

1. [Consume body](https://fetch.spec.whatwg.org/#concept-body-consume-body) calls [fully read body as promise](https://fetch.spec.whatwg.org/#fully-reading-body-as-promise) in step 3.
2. [Fully read body as promise](https://fetch.spec.whatwg.org/#fully-reading-body-as-promise) calls [getting a reader](https://streams.spec.whatwg.org/#readablestream-get-a-reader) in step 1.
3. [Getting a reader](https://streams.spec.whatwg.org/#readablestream-get-a-reader) calls [AcquireReadableStreamDefaultReader](https://streams.spec.whatwg.org/#acquire-readable-stream-reader)
4. [AcquireReadableStreamDefaultReader](https://streams.spec.whatwg.org/#acquire-readable-stream-reader) calls [SetUpReadableStreamDefaultReader](https://streams.spec.whatwg.org/#set-up-readable-stream-default-reader) in step 2.
5. [SetUpReadableStreamDefaultReader](https://streams.spec.whatwg.org/#set-up-readable-stream-default-reader) calls [ReadableStreamReaderGenericInitialize](https://streams.spec.whatwg.org/#readable-stream-reader-generic-initialize) in step 2.
6. [ReadableStreamReaderGenericInitialize](https://streams.spec.whatwg.org/#readable-stream-reader-generic-initialize) sets `stream.[[reader]]` in step 2.

Having a reader means the stream is [locked](https://streams.spec.whatwg.org/#lock).